### PR TITLE
lightning: improve TotalMemoryConsume for write step

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #53858

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by lance6716** [comment]
blocked by known issues

File: pkg/ddl/index.go [hunk 1]
```diff
@@ -1681,7 +1681,8 @@ func getLocalWriterConfig(indexCnt, writerCnt int) *backend.LocalWriterConfig {
 		return writerCfg
 	}
 
-	availMem := memRoot.MaxMemoryQuota() - memRoot.CurrentUsage()
+	// leave some room for objects overhead
+	availMem := memRoot.MaxMemoryQuota() - memRoot.CurrentUsage() - int64(10*size.MB)
```
Comment (D3Hunter) [inline pkg/ddl/index.go:1685]: what this 10M mean?

**Review by D3Hunter — COMMENTED** [review]
will review later

File: pkg/ddl/index.go (hunk 1 above)
Comment (lance6716) [reply pkg/ddl/index.go:1685]: Other place will also call `Consume`, I want to leave some room for them

https://github.com/pingcap/tidb/blob/72d22d60fca2b3d0fd490e5b5ccac9bd56165e36/pkg/ddl/ingest/engine.go#L206

But now I think these call with struct size should be negligible, I will delete them in next commit.

Comment (lance6716) [reply pkg/ddl/index.go:1685]: I see these struct can have many fields, so maybe they will not be negligible in future. I don't want to delete them now.

**Review by D3Hunter — APPROVED** [review]

File: pkg/lightning/backend/local/engine.go [hunk 2]
```diff
@@ -272,6 +275,11 @@ func (e *Engine) TotalMemorySize() int64 {
 			memSize += w.kvBuffer.TotalSize()
 			w.Unlock()
 		}
+		if w.writeBatch != nil {
```
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/lightning/backend/local/engine.go]: https://github.com/pingcap/tidb/pull/53856/commits/6a526e9f1da532c0e0e8012252b959d94e478398

**Review by tangenta — APPROVED** [review]

**Pull request merged**